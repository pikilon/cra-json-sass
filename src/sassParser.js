
const TAB = ' ';
const SIMPLE_TYPES = ['string', 'number', 'boolean']

const sassParser = (json, level = 0, isLast = false, indent = '') => {
  if (typeof json === 'function') return 'THE FUNCTIONS ARE NOT ALLOWED'
  const isSimpleOrFalsy = SIMPLE_TYPES.indexOf(typeof json) >= 0 || !json
  if (isSimpleOrFalsy) return json;

  //is Object
  const finalIndent = indent + TAB
  const isDeepLevel = level >= 1;
  const joinerString = isDeepLevel ? `,\n${finalIndent}` : ';\n$'
  const arrayToIterate = Object.keys(json);

  const lines = arrayToIterate.map((key, index) => {
    const isLastElement = (index + 1) === arrayToIterate.length
    return `${key}: ${sassParser(json[key], level + 1, isLastElement, finalIndent)}`
  })

  const joinedLines = lines.join(joinerString)
  const result = isDeepLevel ? `(\n${finalIndent}${joinedLines}\n${indent})` : `$${joinedLines};\n`
  return result;
}

module.exports = sassParser
