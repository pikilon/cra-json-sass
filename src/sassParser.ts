const TAB = ' ';
const SIMPLE_TYPES = ['string', 'number', 'boolean']

export const sassParser = (json: JsonType, level = 0, indent = '') => {
  const isSimpleOrFalsy = SIMPLE_TYPES.indexOf(typeof json) >= 0
  if (isSimpleOrFalsy) return json;

  //is Object
  const finalIndent = indent + TAB
  const isDeepLevel = level >= 1;
  const joinerString = isDeepLevel ? `,\n${finalIndent}` : ';\n$'
  const arrayToIterate = Object.keys(json)
  const isArray = Array.isArray(json)

  const lines = arrayToIterate.map(key => {
    const value = json[key] as JsonType
    const prefixKey = isArray ? '' : `${key}: `
    const resultString = prefixKey + sassParser(value, level + 1, finalIndent)
    return resultString
  })

  const joinedLines = lines.join(joinerString)
  const result = isDeepLevel ? `(\n${finalIndent}${joinedLines}\n${indent})` : `$${joinedLines};\n`
  return result;
}

