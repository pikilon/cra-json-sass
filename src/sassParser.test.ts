import { sassParser } from './sassParser'
import json from '../tests/basicSample.scss.json';


describe('Sass Parser', () => {
  const result = sassParser(json)

  it('match snapshot', () => {
    expect(result).toMatchSnapshot()
  })

})
