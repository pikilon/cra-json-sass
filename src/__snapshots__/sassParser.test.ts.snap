// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sass Parser match snapshot 1`] = `
"$Button: Button;
$Button__text: Button__text;
$Button--primary: Button--primary;
$isActive: is-active;
$animationSpeed: 0.2;
$backgroundColor: green;
$nested: (
  color: red,
  re-nested: (
   last-value: orange
  )
 );
$array: (
  blue,
  (
   background-color: red,
   font-size: 1rem
  )
 );
"
`;
